### ISSUE TITLE
Show Verified Dice Bug

### ISSUE THREAD

**Issue by General-Dru-Zod** [triplea-game/triplea#3220]
### Engine version
9464

### My Operating System
windows 10

### Map name
World War II 1941 V3

### Can you describe how to trigger the error? (eg: what sequence of actions will recreate it?)
Was in a lobby bot game yesterday playing 1v1. The option "show verified dice" was not showing anything throughout game. Not even "warming up crypto random source".

### Do you have the exact error text? Please copy/paste if so
No errors. Just was checking this option throughout as I was getting some cold dice and opponent was rolling a lot of reds.

### Instead of this error, what should have happened?
Should list something like the following.

Warming up crypto random source. 
List of all rolls by nation, units rolling, location of battle and round of battle.

### Any additional information that may help
I loaded same game save in another bot as a solo test. The "show verified dice" functions normally.

I have noticed this in past but not sure what this is indicating. Is it a dice hack or a bug. Most likely the latter. I have heard from at least one other person that in a bot game this happens.

Here's game save.
[ww2.v3_r4_british._03.02.2018.zip](https://github.com/triplea-game/triplea/files/1777859/ww2.v3_r4_british._03.02.2018.zip)

**Comment by ssoloff** [comment]
Thanks for the heads-up, @General-Dru-Zod.

It looks like the underlying feature is working correctly, but the window is not being updated after it is opened as new dice rolls are generated.  If you close the window and reopen it, you'll see all the dice that were rolled since you previously opened it.

I verified this window has the same incorrect behavior in 3635, so this problem may have been around for a while.

I'll take a look and see if I can figure out what's preventing the window from receiving updates while it's open.

**Comment by General-Dru-Zod** [comment]
@ssoloff It was blank every time I tried. It never listed the info on multiple tries in that game.

**Comment by General-Dru-Zod** [comment]
Just so I know, what is the expected behavior of a dice hack?

**Comment by ssoloff** [comment]
@General-Dru-Zod 

> It was blank every time I tried. It never listed the info on multiple tries in that game.

Got it.  In fact, I may have jumped the gun a bit.  The "real-time" update I was expecting appears to have never been a part of this code.  It's always behaved the way I described.

So now it comes down to figuring out why your client never receives the verified dice rolls to display in the first place.  I'll keep looking.

> Just so I know, what is the expected behavior of a dice hack?

You should see one of the following messages in the console:

* `Server did not unlock random numbers, cheating is suspected`
* `Could not unlock numbers, cheating suspected`

There may be other side effects of a hack, but those were the most obvious in the code.  If I find any more, I'll let you know.

**Comment by ssoloff** [comment]
Ok, I think I know what's going on.  I set up the following scenario:

* Start a lobby with one bot.
* Start a client, connect to the lobby, join the bot, and choose some subset of players.
* Start a second client, connect to the lobby, join the bot, and choose the remaining players.
* Wait for game to start on both clients.
* Open Show Verified Dice on first client; observe no entries.
* Open Show Verified Dice on second client; observe expected entries.

This seems to be the code causing the problem:

https://github.com/triplea-game/triplea/blob/2df5c1f7f656cfba991c30fe0a43a8f607af9f4a/game-core/src/main/java/games/strategy/engine/framework/startup/launcher/ServerLauncher.java#L140-L148

Basically, the bot chooses _only one client_ with which to share verified dice rolls.  I'm not even sure which client it chooses is deterministic.  Therefore, if multiple players are connected to the bot (which is usually the case), only one of them will be able to view verified dice rolls.

I'm not sure if the server code has always behaved this way.  It may be based on some legacy assumption that there would only ever be two nodes in a network game.  I'll need to do further research.

**Comment by General-Dru-Zod** [comment]
A related observation. Privately hosted games do not have this feature available. It would be cool to have something activated optionally, maybe via marti but live.

**Comment by ssoloff** [comment]
@General-Dru-Zod The **Show Verified Dice** command is not available to the player hosting the game, but it should be available to at least one of the players joining the game.  I just checked with 9560 and confirmed the player joining the private host has that command available.  Are you not seeing the same thing?

**Comment by General-Dru-Zod** [comment]
Not sure, I need someone to host a game and try it. I know if I'm just in private host as a non player, the it is available, however always blank.

**Comment by ssoloff** [comment]
@General-Dru-Zod Ah, you're talking about joining an in-progress game and just observing, right?  Then, yeah, you won't ever see any dice rolls in that case.  It's due to the fact that the server picks the one player who will receive the dice rolls when the game is started.  Since you weren't connected to the host when the game started, you can't possibly be picked. :smile:

**Comment by DanVanAtta** [comment]
I don't have full context yet on this, but sounds like this is optimized for the 1v1 case. I would not be very surprised if a multi-verified dice case was ignored or not considered. Most tourneys are 1v1, multiplayer saves game are well known to be near impossible to save and continue (very tough to get 3 schedules to line up let alone 2).

**Comment by ssoloff** [comment]
> sounds like this is optimized for the 1v1 case. I would not be very surprised if a multi-verified dice case was ignored or not considered. Most tourneys are 1v1

Are tournaments usually played through a bot?  If so, this might still be a hot issue if only one of the players can verify dice rolls.  Because, as mentioned above, a 1v1 game via a bot is not really 1v1 from the engine's perspective; rather, it's viewed as a 3-player game.

**Comment by General-Dru-Zod** [comment]
@ssoloff Ok, so in a private host, 3 active players in game, the show verified dice works as follows. For the host player is not functional, not even warming up crypto. For other clients (2 others actively in game) functional, see dice rolls. For guest observer, nothing.

**Comment by ssoloff** [comment]
@General-Dru-Zod Hmm, I'm not seeing the exact same behavior.  In my test, only one client can see dice rolls:

![show-verified-dice-private-host-annotated](https://user-images.githubusercontent.com/4826349/36954587-119001a8-1ff1-11e8-85e7-5fac82d21d00.png)

**Comment by General-Dru-Zod** [comment]
I asked the people in that were actively in the game and they said yes. Unless they completely misunderstood what I was asking them.

**Comment by ssoloff** [comment]
> I will ask again on next available opportunity to confirm.

@General-Dru-Zod Thanks.  Asking them for a screenshot will remove any ambiguity. :smile:
